fix:[CORE-1893] Add NPE check for optional fields in Request/Deny exemption

ershad-paladin opened the pull request:
## Description

- handled NPE for optional fields during Request/Deny exemption
- For Revoke request and Deny exemption, the reason and expiry date are null

### Problem
NPE when optional fields are null

### Solution
handled NPE checks

## Fixes # (issue if any)
NPE

## Type of change

**Please delete options that are not relevant.**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Chore (no code changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also
list any relevant details for your test configuration

- [x] Login as a read-only user and request exemption for a violation
- [x] Revoke the request using the same read-only user. This should revoke the request successfully
- [x] request exemption for a violation as a read-only again.
- [x] Login as admin user and deny the exemption request. This should deny the request successfully

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My commit message/PR follows the contribution guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## **Other Information**:

## List any documentation updates that are needed for the Wiki

plyubich commented on api/pacman-api-compliance/src/main/java/com/tmobile/pacman/api/compliance/repository/ComplianceRepositoryImpl.java:
```diff
@@ -2342,9 +2344,13 @@ public IssueExceptionResponse exemptAndUpdateMultipleIssueDetails(String assetGr
-                            optionalAuditFields.put(Constants.ISSUE_EXEMPTION_REASON, issuesException.getExceptionReason());
+                            Map<String, String> optionalAuditFields = Collections.emptyMap();
+                            if (issuesException.getExceptionEndDate() != null) {
+                                optionalAuditFields.put(Constants.ISSUE_EXEMPTION_EXPIRY_DATE, sdf.format(issuesException.getExceptionEndDate()));
```
If you omit this when it comes empty, what is the consequence? Who is sending this field? If it's the UI, then let's see why it comes empty.

plyubich replied:
@ershad-paladin - was thinking about this over night and here is what I believe is the better approach as this seems more of a hack then a permanent fix.

Let's do a minor refactor to separate (admin vs user functionalities) or (request vs revoke, deny).

1. This is a much cleaner a more robust Admin functionality
2. Avoids confusing check and lets the next dev know exactly what the request for Admin vs User functionality is
3. You can easily abstract this to some base user model class and apply some inheritance.

plyubich commented on api/pacman-api-compliance/src/main/java/com/tmobile/pacman/api/compliance/repository/ComplianceRepositoryImpl.java:
```diff
@@ -2342,9 +2344,13 @@ public IssueExceptionResponse exemptAndUpdateMultipleIssueDetails(String assetGr
+                                optionalAuditFields.put(Constants.ISSUE_EXEMPTION_EXPIRY_DATE, sdf.format(issuesException.getExceptionEndDate()));
+                            }
+                            if (isNotBlank(issuesException.getExceptionReason())) {
+                                optionalAuditFields.put(Constants.ISSUE_EXEMPTION_REASON, issuesException.getExceptionReason());
```
Same here, I believe reason and expiration date are mandatory fields on the UI. If they are, you might not be able to create an exemption.

plyubich reviewed (commented):
See comments

plyubich commented:
Also, how was this working before? Weren't we able to deny and revoke previously?

plyubich reviewed (changes requested):
@ershad-paladin please let me know if you need to discuss this. I believe doing a minor refactoring will greatly improve the quality and readability of the code.

ershad-paladin commented:
> @ershad-paladin please let me know if you need to discuss this. I believe doing a minor refactoring will greatly improve the quality and readability of the code.

for admins we have the APIs /api/compliance/v2/issue/add-exception (called when admin adds an exemption)
/api/compliance/v2/issue/revoke-exception  (called when admin revokes the exemption)

For all user actions related to request/revoke/deny the API /api/compliance/v2/issue/create-revoke-user-exemption is called. We made changes in the common method which has separate logical block for each user action keeping the common logic.

As for refactoring, I have added the required changes in the correct logical block based on user actions. I have added the exemption reason and expiry date only when user requests and exemption or the exemption is approved, because those are the only cases when we want to populate this data.

Please review and let me know if any.
